### ISSUE TITLE
Testing UNIQUE Constraint

### ISSUE THREAD

**Issue by healiseu** [josiahcarlson/rom#118]
Hi Josiah,
first let me thank you for sharing freely and openly `rom` library. Someone like me can definitely learn many things by studying your code. Although I do have significant programming experience in several languages and a computer science background, I consider myself to be a newcomer in the world of Redis and certainly I am not an expert in Python.

I have decided to look at your code because there is a chance to use it in developing further [TRIADB][1] project. In fact I am confident that any object-persistence, object-relational project is highly correlated. Hope we find some time to discuss about this, in case you find TRIADB approach interesting.

This is enough for introduction. Regarding to testing features of your toolkit, I have started writing small programs and I plan to make them also available for your users. One of them is the following, note that I am using `rom` v0.41.2 and Python 3.6

```python

from rom.util import session
from rom.model import Model
from rom.columns import Integer, Text

class UndxModel(Model):
    txtkey = Text(required=True, unique=True)
    intkey = Integer(unique=True)

# Try to create an instance with an attribute that is required but missing
try:
    item = UndxModel()
except Exception as ex:
    print(ex)

# Try to create an instance with an attribute that is not present in the model
# This is NOT raising an exception at this point
try:
    item1 = UndxModel(txtkey='hello', foo='bar')
except Exception as ex:
    print(ex)

# The following raises an exception, object has no attribute foo
try:
    item1.foo
except Exception as ex:
    print(ex)

# Save the instance
item1.save()

# Result of to_dict : {'id': 1, 'intkey': None, 'txtkey': 'hello'} 
# Notice the None value
item1.to_dict()

# Attempt to create a second instance with 'txtkey' : None
# This also raises an exception due to required constraint: 
# UndxModel.txtkey cannot be missing
try:
    item2 = UndxModel(intkey=20)
except Exception as ex:
    print(ex)

# Call .save() on all modified entities in the session
session.commit()

UndxModel.query.select('id', 'txtkey', 'intkey').all()
```
So it seems exceptions in general  are handled pretty well. My question is whether you should also add another one if the user tries to create an instance with an attribute that is not a member of the object class. 

I also have another question. I noticed that in Redis you are using a hash table structure to create an index for each unique class attribute and you store there all the unique values. Is that more efficient than writing string value keys ? Would you think SETNX could possibly make a difference in performance here ?

The reason I am asking this is because in TRIADB, **value domain of primitive data types is separated and differentiated from the set of unique references to values that the type of a class attribute takes**. Everything, including the Hypergraph mechanism, is based on single instances.  Since this procedure is **heavily repeated** and it is fundamental in the way data are stored and retrieved I am looking into ways to optimize this part. In previous implementations it took a really long time (**min**s) for a small volume data (**MB**s) during the ingestion stage. So, in massive insertion, I need something really fast, e.g. disabling perhaps caching or other control features of rom.

[1]: http://healis.eu/triadb/

**Comment by josiahcarlson** [comment]
> Hi Josiah,
> first let me thank you for sharing freely and openly rom library. Someone like me can definitely learn many things by studying your code. Although I do have significant programming experience in several languages and a computer science background, I consider myself to be a newcomer in the world of Redis and certainly I am not an expert in Python.

Happy to help, we all start somewhere. :)

> I have decided to look at your code because there is a chance to use it in developing further TRIADB project. In fact I am confident that any object-persistence, object-relational project is highly correlated. Hope we find some time to discuss about this, in case you find TRIADB approach interesting.

Definitely. The only thing missing from rom is joins and aggregates. Other than that, it offers about 95% of the day-to-day functionality of a project like SQLAlchemy. That's on purpose. Only thing stopping rom from being more is me. I'm not interested in spending the time on rom to make it more, because so far, it hasn't really been a money maker, and I need to feed my family.

> This is enough for introduction. Regarding to testing features of your toolkit, I have started writing small programs and I plan to make them also available for your users. One of them is the following, note that I am using rom v0.41.2 and Python 3.6

How are you planning on making them available?

>     from rom.util import session
>     from rom.model import Model
>     from rom.columns import Integer, Text
> 
>     class UndxModel(Model):
>         txtkey = Text(required=True, unique=True)
>         intkey = Integer(unique=True)
> 
>     # Try to create an instance with an attribute that is required but missing
>     try:
>         item = UndxModel()
>     except Exception as ex:
>         print(ex)
> 
>     # Try to create an instance with an attribute that is not present in the model
>     # This is NOT raising an exception at this point
>     try:
>         item1 = UndxModel(txtkey='hello', foo='bar')
>     except Exception as ex:
>         print(ex)

Until I executed it, I wouldn't have believed that this wouldn't have raised an error. I could have sworn that we used to raise an exception on extra attributes. But no, looking at the `Model.__init__` code, there isn't even a check for extra attributes. We just drop them on the ground if they are passed.

>     # The following raises an exception, object has no attribute foo
>     try:
>         item1.foo
>     except Exception as ex:
>         print(ex)

This exception is not surprising, given the previous non-exception, and that `foo` isn't supposed to be there.

>     # Save the instance
>     item1.save()
>     
>     # Result of to_dict : {'id': 1, 'intkey': None, 'txtkey': 'hello'} 
>     # Notice the None value
>     item1.to_dict()

Yep. Non-values are encoded as `None` in the dict, and are not passed into Redis (generally).

>     # Attempt to create a second instance with 'txtkey' : None
>     # This also raises an exception due to required constraint: 
>     # UndxModel.txtkey cannot be missing
>     try:
>         item2 = UndxModel(intkey=20)
>     except Exception as ex:
>         print(ex)
> 
>     # Call .save() on all modified entities in the session
>     session.commit()
> 
>     UndxModel.query.select('id', 'txtkey', 'intkey').all()

This returns a `False` value for intkey, which is weird. Should be `None`.

> So it seems exceptions in general are handled pretty well. My question is whether you should also add another one if the user tries to create an instance with an attribute that is not a member of the object class.

Yeah, an exception should be raised. Not having it raise an exception, but also not storing it violates the Python zens of "Errors should never pass silently." and "Unless explicitly silenced." I think this was meant to handle the case of data already in Redis where you are overlaying rom on top. I'll handle that with an explicit check for whether this is a new object, or being loaded, and an optional flag of `_extra_ok`, also in the exception message.

While looking into this, I also noticed that some weird session stuff was going on since I fixed bugs #108 and #109, so I've also got a fix for that too. Thank you for getting me to look into this.

> I also have another question. I noticed that in Redis you are using a hash table structure to create an index for each unique class attribute and you store there all the unique values. Is that more efficient than writing string value keys ? Would you think SETNX could possibly make a difference in performance here ?

Generally, yes, it is more efficient. Whatever inefficiencies you lose on by having another hash table (if it gets big), you more than make up for by removing the `<TableName>:<ColumnName>:uidx:` prefix that would be necessary if it was all in the main keyspace. So call it at least a 16 byte/entry savings, at minimum. For small unique indexes, all of the data is stored packed in a small string in Redis, so you do even better.

Also, Lua's a big part of how/why rom works, and we've got to make a trip through Lua in Redis. Using `SETNX` doesn't do us any good, unless we are trying to avoid Lua, explicitly avoid using Redis' rich types, etc. And ... that is not a design goal of rom; the "Redis Object Mapper".

> The reason I am asking this is because in TRIADB, value domain of primitive data types is separated and differentiated from the set of unique references to values that the type of a class attribute takes. Everything, including the Hypergraph mechanism, is based on single instances.

I'm not sure I entirely understood that. But guessing here, do you mean that in TRIADB, because objects are stored as hashes when using a Redis backend, unique indexes can't also be stored as hashes?

> Since this procedure is heavily repeated and it is fundamental in the way data are stored and retrieved I am looking into ways to optimize this part. In previous implementations it took a really long time (mins) for a small volume data (MBs) during the ingestion stage. So, in massive insertion, I need something really fast, e.g. disabling perhaps caching or other control features of rom.

We improved item deletion in #117 (removing data can't induce a unique key violation / data error if you don't allow models with foreign key relationships), but item creation may be a bit funky (and may require not having foreign key relationships). I think I've got most of the details reasonably worked out in my head, but I'll have to think it over for a while before I move forward with my fix.

Speaking of which... the stuff I mentioned above as being broken/weird is fixed in 0.42.0, which I just released.

**Comment by healiseu** [comment]
Hi @josiahcarlson , thank you for the efforts to fix the bug and for replying to my question.

As I mentioned some time ago I have decided to test ROM. The time is now and I have written my first coding example that tests ROM column primitive types. You may download my [jupyter notebook][2] from the [library I created for your project][1] at Microsoft Azure notebooks cloud

[1]: https://notebooks.azure.com/athanassios/libraries/redisrom
[2]: https://notebooks.azure.com/athanassios/libraries/redisrom/html/ROM%20Testing%20column%20types%20and%20indexes.ipynb
